jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUTreeIsomorphismInspector.java

```diff
@@ -0,0 +1,349 @@
+        this.backwardMapping = new HashMap<>();
+
+        // Are both graphs empty?
+        if (Objects.isNull(root1) && Objects.isNull(root2))
+            return true;
+
+        @SuppressWarnings("unchecked")
+        Map<V, Integer>[] canonicalName = new Map[2];
+        canonicalName[0] = new HashMap<>(tree1.vertexSet().size());
+        canonicalName[1] = new HashMap<>(tree2.vertexSet().size());
+
+        List<List<V>> nodesByLevel1 = computeLevels(tree1, root1);
+        List<List<V>> nodesByLevel2 = computeLevels(tree2, root2);
+
+        if (nodesByLevel1.size() != nodesByLevel2.size())
+            return false;
+
+        final int MAX_LEVEL = nodesByLevel1.size() - 1;
+
+        @SuppressWarnings("unchecked")
+        Map<String, Integer> canonicalNameToInt = new HashMap<>();
+
+        List<Map<Integer, List<V>>> sameLabelBagsByLevel = new ArrayList<>(nodesByLevel1.size());
+
+        int freshName = 0;
+
+        for (int lvl = MAX_LEVEL; lvl >= 0; lvl--) {
+            @SuppressWarnings("unchecked")
+            List<V>[] level = new List[2];
+
+            level[0] = nodesByLevel1.get(lvl);
+            level[1] = nodesByLevel2.get(lvl);
+
+            if (level[0].size() != level[1].size()) {
+                return false;
+            }
+
+            final int n = level[0].size();
+
+            @SuppressWarnings("unchecked")
+            Map<Integer, List<V>> sameLabelBags = new HashMap<>(n);
+
+            for (int k = 0; k < 2; k++) {
+                for (int i = 0; i < n; i++) {
+                    V u = level[k].get(i);
+
+                    Graph<V, E> graph = tree1;
+
+                    if (k == 1)
+                        graph = tree2;
+
+                    List<Integer> list = new ArrayList<>();
+                    for (E edge: graph.edgesOf(u)){
+                        V v = Graphs.getOppositeVertex(graph, edge, u);
+                        int name = canonicalName[k].getOrDefault(v, -1);
+
+                        if (name != -1){
+                            list.add(name);
+                        }
+                    }
+
+                    RadixSort.sort(list);
+
+                    StringBuilder sb = new StringBuilder();
+                    sb.append(1).append(",");
+
+                    for (int x: list)
+                        sb.append(x).append(",");
+
+                    sb.append(0);
+
+                    String name = sb.toString();
+                    Integer intName = canonicalNameToInt.get(name);
+
+                    if (intName == null){
+                        canonicalNameToInt.put(name, freshName);
+                        intName = freshName;
+                        freshName++;
+                    }
+
+                    canonicalName[k].put(u, intName);
+
+                    if (k == 1){
+                        List<V> bag = sameLabelBags.computeIfAbsent(intName, x -> new ArrayList<>());
+                        bag.add(u);
```

Comment:
Does this do anything?

Reply:
It should add `u` fo the bag associated with `intName`

Reply:
ok, better to write it in one line 
```
sameLabelBags.computeIfAbsent(intName, x -> new ArrayList<>()).add(u);
```